### PR TITLE
Add ViLT

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [ViLT](https://arxiv.org/abs/2102.03334) (Vision and Language Transformer).

It's a very nice, minimal multi-modal model, as it only adds a text embedding layer to an existing ViT.

I've defined the following head models:

* `ViltForMaskedLM`
* `ViltForVisualQuestionAnswering`
* `ViltForNaturalLanguageVisualReasoning`
* `ViltForImageRetrievalTextRetrieval` (CLIP-like model).

To do:

- [x] add `ViltForNaturalLanguageVisualReasoning` to the tests. However, I do have a question here: it's the only model that requires `config.modality_type_vocab_size = 3` instead of 2. How can I handle this exception in the tests? I could do it like this:

```
for model_class in self.all_model_classes:
    config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
    config.return_dict = True

    if model_class.__name__ == "ViltForNaturalLanguageVisualReasoning":
       config.modality_type_vocab_size = 3
```
But that's not ideal as it would require overwrite each individual test.

Update: fixed by create a separate `ModelTester` for this particular model, that overrides the `get_config`.

### CODE REVIEW

File: docs/source/index.rst [hunk 1]
```diff
@@ -0,0 +1,720 @@
+Transformers
```
Comment (LysandreJik) [inline docs/source/index.rst:1]: I think this can be removed

File: docs/source/model_doc/vilt.mdx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# ViLT
+
+## Overview
+
+The VilT model was proposed in [ViLT: Vision-and-Language Transformer Without Convolution or Region Supervision](https://arxiv.org/abs/2102.03334)
```
Comment (LysandreJik) [inline docs/source/model_doc/vilt.mdx:17]:
```suggestion
The ViLT model was proposed in [ViLT: Vision-and-Language Transformer Without Convolution or Region Supervision](https://arxiv.org/abs/2102.03334)
```

File: docs/source/model_doc/vilt.mdx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+find it problematic in terms of both (1) efficiency/speed, that simply extracting input features requires much more
+computation than the multimodal interaction steps; and (2) expressive power, as it is upper bounded to the expressive
+power of the visual embedder and its predefined visual vocabulary. In this paper, we present a minimal VLP model,
+Vision-and-Language Transformer (ViLT), monolithic in the sense that the processing of visual inputs is drastically
+simplified to just the same convolution-free manner that we process textual inputs. We show that ViLT is up to tens of
+times faster than previous VLP models, yet with competitive or better downstream task performance.*
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
```
Comment (LysandreJik) [inline docs/source/model_doc/vilt.mdx:35]: This should be completed

File: docs/source/model_doc/vilt.mdx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+This model was contributed by [nielsr](https://huggingface.co/nielsr). The original code can be found [here](https://github.com/dandelin/ViLT).
+
+## ViltConfig
+
+[[autodoc]] ViltConfig
+    - all
```
Comment (LysandreJik) [inline docs/source/model_doc/vilt.mdx:42]: No need for `- all`
```suggestion
```

File: src/transformers/models/vilt/configuration_vilt.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```
Comment (LysandreJik) [inline src/transformers/models/vilt/configuration_vilt.py:2]:
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
:boom:

File: src/transformers/models/vilt/modeling_vilt.py [hunk 6]
```diff
@@ -0,0 +1,1296 @@
...
+            embed_dim=config.hidden_size,
+        )
+        num_patches = self.patch_embeddings.num_patches
+        self.position_embeddings = nn.Parameter(torch.zeros(1, num_patches + 1, config.hidden_size))
+        # modality type (text/patch) embeddings
+        self.token_type_embeddings = nn.Embedding(config.modality_type_vocab_size, config.hidden_size)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.config = config
+
+    def visual_embed(self, pixel_values, pixel_mask, max_image_length=200):
```
Comment (LysandreJik) [inline src/transformers/models/vilt/modeling_vilt.py:88]: Could you explain the code a bit? Through longer variable names and comments, as it is otherwise quite hard to follow. Thanks!

File: src/transformers/models/vilt/modeling_vilt.py [hunk 7]
```diff
@@ -0,0 +1,1296 @@
...
+        self.num_patches = num_patches
+
+        self.projection = nn.Conv2d(num_channels, embed_dim, kernel_size=patch_size, stride=patch_size)
+
+    def forward(self, pixel_values):
+        batch_size, num_channels, height, width = pixel_values.shape
+        # if height != self.image_size[0] or width != self.image_size[1]:
+        #     raise ValueError(
+        #         f"Input image size ({height}*{width}) doesn't match model ({self.image_size[0]}*{self.image_size[1]})."
+        #     )
```
Comment (LysandreJik) [inline src/transformers/models/vilt/modeling_vilt.py:307]: If unnecessary you can remove this

File: src/transformers/models/vilt/modeling_vilt.py [hunk 8]
```diff
@@ -0,0 +1,1296 @@
...
+        x = self.projection(pixel_values)
+        return x
+
+
+class ViltSelfAttention(nn.Module):
```
Comment (LysandreJik) [inline src/transformers/models/vilt/modeling_vilt.py:312]: This looks very similar to other `SelfAttention` classes - could it use a `# Copied from` statement?

File: src/transformers/models/vilt/modeling_vilt.py [hunk 9]
```diff
@@ -0,0 +1,1296 @@
...
+    layernorm applied before each block.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+    def forward(self, hidden_states, input_tensor):
+
```
Comment (LysandreJik) [inline src/transformers/models/vilt/modeling_vilt.py:384]:
```suggestion
```

File: src/transformers/models/vilt/modeling_vilt.py [hunk 10]
```diff
@@ -0,0 +1,1296 @@
...
+        return MaskedLMOutput(
+            loss=masked_lm_loss,
+            logits=mlm_logits,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+
+class ViltITMHead(nn.Module):
```
Comment (LysandreJik) [inline src/transformers/models/vilt/modeling_vilt.py:911]: Should this be used? It isn't used anywhere

Comment (LysandreJik) [reply src/transformers/models/vilt/modeling_vilt.py:911]: Also the checkpoints indicate `itm` so I suppose this is important :)

File: utils/check_repo.py [hunk 11]
```diff
@@ -102,6 +102,9 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = PRIVATE_MODELS.copy() + [
     # models to ignore for model xxx mapping
+    "ViltForMaskedLM",
```
Comment (LysandreJik) [inline utils/check_repo.py:105]: `AutoModelForMaskeLM` should correctly return this, no?

File: tests/test_modeling_vilt.py [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+            out_2 = out_2[~np.isnan(out_2)]
+            max_diff = np.amax(np.abs(out_1 - out_2))
+            self.assertLessEqual(max_diff, 1e-5)
+
+    @unittest.skip(
+        reason="""VilT samples image tokens from a multinomial distribution, resulting in not deterministic
+                            hidden states"""
+    )
+    def test_model_outputs_equivalence(self):
+        pass
```
Comment (LysandreJik) [inline tests/test_modeling_vilt.py:348]: :+1:

**Review by LysandreJik — COMMENTED** [review]
This looks good, thanks for working on it @NielsRogge!

I left a few comments, and would love @sgugger's review before this is merged.

**Comment by sgugger** [comment]
I'd like the PR to be green (or mostly green) before reviewing.

File: utils/check_repo.py (hunk 11 above)
Comment (NielsRogge) [reply utils/check_repo.py:106]: I asked @Narsil about this, but the `AutoModelForMaskedLM` doesn't currently accept models that take several modalities as input.

Comment (Narsil) [reply utils/check_repo.py:106]: The pipeline `fill-mask` doesn't work with multi modalities. And it assumes every `AutoModelForMaskedLM` is for filling text only masks.
As I mentionned too:

- Either we don't make it `AutoMaskedLM`
- Or we make it `AutoMaskedLM` but we need an escape hatch of some kind so that the pipeline can know it's not supposed to work.
- `AutoModel` should work nonetheless (I assume this discards that)

**Comment by NielsRogge** [comment]
@sgugger should be mostly green now.

File: docs/source/_toctree.yml [hunk 13]
```diff
@@ -280,7 +280,9 @@
       title: UniSpeech
     - local: model_doc/unispeech-sat
       title: UniSpeech-SAT
-    - local: model_doc/vision-encoder-decoder
+    - local: model_doc/vilt
+      title: ViLT
+    - local: model_doc/visionencoderdecoder
```
Comment (sgugger) [inline docs/source/_toctree.yml:285]: This line should not be changed. Each model doc page should have the same name as the model type.

File: docs/source/_toctree.yml [hunk 14]
```diff
@@ -280,7 +280,9 @@
       title: UniSpeech
     - local: model_doc/unispeech-sat
       title: UniSpeech-SAT
-    - local: model_doc/vision-encoder-decoder
+    - local: model_doc/vilt
```
Comment (sgugger) [inline docs/source/_toctree.yml:283]: Also remember to add the model to the main README.

File: docs/source/model_doc/vilt.mdx [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+[[autodoc]] ViltModel
+    - forward
+
+## ViltForMaskedLM
+
+[[autodoc]] ViltForMaskedLM
+    - forward
+
+## ViltForVisualQuestionAnswering
```
Comment (sgugger) [inline docs/source/model_doc/vilt.mdx:69]: I would remove the visual here. We have dropped the audio in the speech for model for sequence classification for instance, so there is no need to put it here IMO.

File: docs/source/model_doc/vilt.mdx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+[[autodoc]] ViltForVisualQuestionAnswering
+    - forward
+
+## ViltForNaturalLanguageVisualReasoning
```
Comment (sgugger) [inline docs/source/model_doc/vilt.mdx:74]: I have no idea what Natural Language Visual Reasoning means, so there is probably a better name to find here.

File: docs/source/model_doc/vilt.mdx [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+[[autodoc]] ViltForNaturalLanguageVisualReasoning
+    - forward
+
+## ViltForImageRetrievalTextRetrieval
```
Comment (sgugger) [inline docs/source/model_doc/vilt.mdx:79]: `ViltForImageAndTextRetrieval` seems clearer to me.

File: src/transformers/__init__.py [hunk 18]
```diff
@@ -2604,6 +2622,7 @@
         from .utils.dummy_pytorch_quantization_and_torch_objects import *
 
     if is_torch_available():
+
```
Comment (sgugger) [inline src/transformers/__init__.py:2625]:
```suggestion
```

File: src/transformers/__init__.py [hunk 19]
```diff
@@ -626,6 +629,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
Comment (sgugger) [inline src/transformers/__init__.py:632]:
```suggestion
```

File: src/transformers/models/auto/configuration_auto.py [hunk 20]
```diff
@@ -192,6 +194,7 @@
 MODEL_NAMES_MAPPING = OrderedDict(
     [
         # Add full (and cased) model names here
+        ("vilt", "Vilt"),
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:197]:
```suggestion
        ("vilt", "ViLT"),
```
It's ViLT, no?

File: src/transformers/models/auto/modeling_auto.py [hunk 21]
```diff
@@ -152,6 +154,7 @@
     [
         # Model with LM heads mapping
         ("nystromformer", "NystromformerForMaskedLM"),
+        ("imagegpt", "ImageGPTForCausalLM"),
```
Comment (sgugger) [inline src/transformers/models/auto/modeling_auto.py:157]: This is unrelated to this new model addition and should be done in its own PR.

File: src/transformers/models/deit/modeling_deit.py [hunk 22]
```diff
@@ -297,12 +297,6 @@ def forward(self, hidden_states, head_mask=None, output_attentions=False):
 
         # in DeiT, layernorm is also applied after self-attention
         layer_output = self.layernorm_after(hidden_states)
-
```
Comment (sgugger) [inline src/transformers/models/deit/modeling_deit.py:300]: Should be done in its own PR too? Unless you're relying on this with some Copied from.

File: src/transformers/models/vilt/__init__.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
+# flake8: noqa
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (sgugger) [inline src/transformers/models/vilt/__init__.py:5]:
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

File: src/transformers/models/vilt/configuration_vilt.py [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+            The epsilon used by the layer normalization layers.
+        image_size (`int`, *optional*, defaults to `384`):
+            The size (resolution) of each image.
+        patch_size (`int`, *optional*, defaults to `32`):
+            The size (resolution) of each patch.
+        num_channels (`int`, *optional*, defaults to `3`):
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+        max_image_length (`int`, *optional*, defaults to `-1`):
```
Comment (sgugger) [inline src/transformers/models/vilt/configuration_vilt.py:77]: The integers should not be in code bocks.

File: src/transformers/models/vilt/convert_vilt_original_to_pytorch.py [hunk 25]
```diff
@@ -0,0 +1,297 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team.
```
Comment (sgugger) [inline src/transformers/models/vilt/convert_vilt_original_to_pytorch.py:2]:
```suggestion
# Copyright 2022 The HuggingFace Inc. team.
```

File: src/transformers/models/vilt/feature_extraction_vilt.py [hunk 26]
```diff
@@ -0,0 +1,292 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```
Comment (sgugger) [inline src/transformers/models/vilt/feature_extraction_vilt.py:2]:
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```

File: src/transformers/models/vilt/modeling_vilt.py [hunk 27]
```diff
@@ -0,0 +1,1268 @@
...
+    def visual_embed(self, pixel_values, pixel_mask, max_image_length=200):
+        _, _, ph, pw = self.patch_embeddings.projection.weight.shape
+
+        x = self.patch_embeddings(pixel_values)
+        x_mask = pixel_mask[:, None, :, :].float()
+        x_mask = nn.functional.interpolate(x_mask, size=(x.shape[2], x.shape[3])).long()
+        x_h = x_mask[:, 0].sum(dim=1)[:, 0]
+        x_w = x_mask[:, 0].sum(dim=2)[:, 0]
+
+        B, C, H, W = x.shape
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:91]: Let's not use one-letter variable names and capitalize variable names that do not represent a class.

File: src/transformers/models/vilt/modeling_vilt.py [hunk 28]
```diff
@@ -0,0 +1,1268 @@
...
+
+        pos_embed = pos_embed.flatten(2).transpose(1, 2)
+        x = x.flatten(2).transpose(1, 2)
+        patch_index = (
+            torch.stack(
+                torch.meshgrid(torch.arange(x_mask.shape[-2]), torch.arange(x_mask.shape[-1]), indexing="ij"),
+                dim=-1,
+            )[None, None, :, :, :]
+            .expand(x_mask.shape[0], x_mask.shape[1], -1, -1, -1)
+            .flatten(1, 3)
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:118]: This should be done in several lines.

File: src/transformers/models/vilt/modeling_vilt.py [hunk 29]
```diff
@@ -0,0 +1,1268 @@
...
+        )
+        x_mask = x_mask.flatten(1)
+
+        if max_image_length < 0 or max_image_length is None or not isinstance(max_image_length, int):
+            # suppose aug is 800 x 1333, then, maximum effective res is 800 x 1333 (if one side gets bigger, the other will be constrained and be shrinked)
+            # (800 // self.patch_size) * (1333 // self.patch_size) is the maximum number of patches that single image can get.
+            # if self.patch_size = 32, 25 * 41 = 1025
+            # if res is 384 x 640, 12 * 20 = 240
+            eff = x_h * x_w
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:127]: What does `eff` min? This should be a more descriptive name.

File: src/transformers/models/vilt/modeling_vilt.py [hunk 30]
```diff
@@ -0,0 +1,1268 @@
...
+            if p <= 0:
+                valid_choice = torch.multinomial(torch.ones(v).float(), max_image_length)
+                select.append(valid_row_idx[i][valid_choice])
+            else:
+                pad_choice = torch.multinomial(torch.ones(nv).float(), p, replacement=True)
+                select.append(
+                    torch.cat(
+                        [valid_row_idx[i], non_valid_row_idx[i][pad_choice]],
+                        dim=0,
+                    )
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:154]:
```suggestion
                    torch.cat([valid_row_idx[i], non_valid_row_idx[i][pad_choice]], dim=0)
```
Fits in one line.

File: src/transformers/models/vilt/modeling_vilt.py [hunk 31]
```diff
@@ -0,0 +1,1268 @@
...
+        text_embeds, image_embeds = (
+            text_embeds
+            + self.token_type_embeddings(
+                torch.zeros_like(attention_mask, dtype=torch.long, device=text_embeds.device)
+            ),
+            image_embeds
+            + self.token_type_embeddings(
+                torch.full_like(image_masks, image_token_type_idx, dtype=torch.long, device=text_embeds.device)
+            ),
+        )
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:210]: This is too complex. `text_embeds` and `image_embeds` should each incremented on their one lines.

File: src/transformers/models/vilt/modeling_vilt.py [hunk 32]
```diff
@@ -0,0 +1,1268 @@
...
+        embeddings = self.dropout(embeddings)
+        return embeddings
+
+
+# Based on timm implementation, which can be found here:
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py
+class PatchEmbeddings(nn.Module):
+    """
+    Image to Patch Embedding.
+
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:282]:
```suggestion
```

File: src/transformers/models/vilt/modeling_vilt.py [hunk 33]
```diff
@@ -0,0 +1,1268 @@
...
+        head_mask=None,
+        inputs_embeds=None,
+        image_embeds=None,
+        labels=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        r"""
+        labels (*torch.FloatTensor* of shape *(batch_size, num_labels)*, *optional*):
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:971]:
```suggestion
        labels (`torch.FloatTensor` of shape `(batch_size, num_labels)`, *optional*):
```

File: src/transformers/models/vilt/modeling_vilt.py [hunk 34]
```diff
@@ -0,0 +1,1268 @@
...
+        head_mask=None,
+        inputs_embeds=None,
+        image_embeds=None,
+        labels=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        r"""
+        labels (*torch.LongTensor* of shape *(batch_size,)*, *optional*):
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:1075]:
```suggestion
        labels (`torch.LongTensor` of shape `(batch_size,)`, *optional*):
```

File: src/transformers/models/vilt/modeling_vilt.py [hunk 35]
```diff
@@ -0,0 +1,1268 @@
...
```
Comment (sgugger) [inline src/transformers/models/vilt/modeling_vilt.py:1184]:
```suggestion
        labels (`torch.LongTensor` of shape `(batch_size,)`, *optional*):
```

File: src/transformers/models/vilt/processing_vilt.py [hunk 36]
```diff
@@ -0,0 +1,173 @@
...
+        return_special_tokens_mask: bool = False,
+        return_offsets_mapping: bool = False,
+        return_length: bool = False,
+        verbose: bool = True,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        **kwargs
+    ) -> BatchEncoding:
+        """
+        This method uses ViltFeatureExtractor's [`~ViltFeatureExtractor.__call__`] method to prepare image(s) for the
+        model, and BertTokenizer's [`~BertTokenizer.__call__`] to prepare text for the model.
```
Comment (sgugger) [inline src/transformers/models/vilt/processing_vilt.py:132]:
```suggestion
        This method uses [`ViltFeatureExtractor.__call__`] method to prepare image(s) for the
        model, and [`BertTokenizer.__call__`] to prepare text for the model.
```

File: src/transformers/models/vit/modeling_vit.py [hunk 37]
```diff
@@ -326,12 +326,6 @@ def forward(self, hidden_states, head_mask=None, output_attentions=False):
 
         # in ViT, layernorm is also applied after self-attention
         layer_output = self.layernorm_after(hidden_states)
-
```
Comment (sgugger) [inline src/transformers/models/vit/modeling_vit.py:329]: Same comment as for Beit above.

File: tests/test_feature_extraction_vilt.py [hunk 38]
```diff
@@ -0,0 +1,247 @@
...
+        image_inputs = prepare_image_inputs(self.feature_extract_tester, equal_resolution=False, torchify=True)
+        for image in image_inputs:
+            self.assertIsInstance(image, torch.Tensor)
+
+        # Test whether the method "pad_and_return_pixel_mask" and calling the feature extractor return the same tensors
+        encoded_images_with_method = feature_extractor_1.pad_and_create_pixel_mask(image_inputs, return_tensors="pt")
+        encoded_images = feature_extractor_2(image_inputs, return_tensors="pt")
+
+        assert torch.allclose(encoded_images_with_method["pixel_values"], encoded_images["pixel_values"], atol=1e-4)
+        assert torch.allclose(encoded_images_with_method["pixel_mask"], encoded_images["pixel_mask"], atol=1e-4)
```
Comment (sgugger) [inline tests/test_feature_extraction_vilt.py:247]: Use `self.assertTrue` here.

File: tests/test_modeling_vilt.py [hunk 39]
```diff
@@ -0,0 +1,549 @@
...
+        config = self.get_config()
+
+        return (
+            config,
+            input_ids,
+            token_type_ids,
+            input_mask,
+            pixel_values,
+            token_labels,
+        )
```
Comment (sgugger) [inline tests/test_modeling_vilt.py:132]: Firs in one line.

File: tests/test_modeling_vilt.py [hunk 40]
```diff
@@ -0,0 +1,549 @@
...
+@require_torch
+class ViltModelTest(ModelTesterMixin, unittest.TestCase):
+
+    all_model_classes = (
+        (
+            ViltModel,
+            ViltForVisualQuestionAnswering,
+            ViltForImageRetrievalTextRetrieval,
+            ViltForMaskedLM,
+            # ViltForNaturalLanguageVisualReasoning,
```
Comment (sgugger) [inline tests/test_modeling_vilt.py:204]: How about making a second tester for this model that inherits from `ViltModelTester` and `ViltModelTest` but overrides the `get_config` method to set the proper variable to 3?

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding this model!

Regarding your question for the tests, I think the easiest would be to define a new `Tester` and `Test` class for `ViltForNaturalLanguageVisualReasoning` that inehrits from the main `Tester` and `Test` class this PR adds, then overrides the method to get the config, so that you don't have to rewrite all the tests.

File: docs/source/model_doc/vilt.mdx (hunk 15 above)
Comment (NielsRogge) [reply docs/source/model_doc/vilt.mdx:69]: Ok, pinging @Narsil here in case we want to make a VQA pipeline.

File: src/transformers/models/auto/modeling_auto.py (hunk 21 above)
Comment (NielsRogge) [reply src/transformers/models/auto/modeling_auto.py:157]: This is a mistake, I've removed it.

File: tests/test_modeling_vilt.py (hunk 40 above)
Comment (NielsRogge) [reply tests/test_modeling_vilt.py:197]: Great tip, thanks for the advice!

File: docs/source/model_doc/vilt.mdx (hunk 15 above)
Comment (sgugger) [reply docs/source/model_doc/vilt.mdx:69]: We can still have a VQA pipeline with the model being named `ViltForQuestionAnswering`. Just look at the the models in the audio classification pipeline ;-)

File: docs/source/model_doc/vilt.mdx (hunk 16 above)
Comment (NielsRogge) [reply docs/source/model_doc/vilt.mdx:74]: It's because this model was fine-tuned on NLVR: https://lil.nlp.cornell.edu/nlvr/

File: src/transformers/models/deit/modeling_deit.py (hunk 22 above)
Comment (NielsRogge) [reply src/transformers/models/deit/modeling_deit.py:300]: Yes it's because of Copied from statements which I added.

File: docs/source/model_doc/vilt.mdx (hunk 15 above)
Comment (NielsRogge) [reply docs/source/model_doc/vilt.mdx]: Ok, renamed it.

File: src/transformers/models/vilt/processing_vilt.py [hunk 41]
```diff
@@ -0,0 +1,175 @@
...
+                - a path to a *directory* containing a feature extractor file saved using the
+                  [`~SequenceFeatureExtractor.save_pretrained`] method, e.g., `./my_model_directory/`.
+                - a path or url to a saved feature extractor JSON *file*, e.g.,
+                  `./my_model_directory/preprocessor_config.json`.
+            **kwargs
+                Additional keyword arguments passed along to both [`SequenceFeatureExtractor`] and
+                [`PreTrainedTokenizer`]
+        """
+        feature_extractor = ViltFeatureExtractor.from_pretrained(pretrained_model_name_or_path, **kwargs)
+        tokenizer = BertTokenizer.from_pretrained(pretrained_model_name_or_path, **kwargs)
```
Comment (NielsRogge) [inline src/transformers/models/vilt/processing_vilt.py:108]: @sgugger as you can see here, I'm instantiating a slow tokenizer from the hub.

However, it may be that people want to use the fast tokenizer instead. Should I add an argument `use_fast` to the `from_pretrained` method here?

Comment (sgugger) [reply src/transformers/models/vilt/processing_vilt.py:108]: Why use the slow class when we have a fast one available? IMO there should be no arg, just `BertTokenizerFast`.

**Comment by sgugger** [comment]
Note: with the new build dev job merged, you can preview the doc [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_14895/en/model_doc/vilt) :-)

File: src/transformers/models/vilt/processing_vilt.py (hunk 41 above)
Comment (NielsRogge) [reply src/transformers/models/vilt/processing_vilt.py]: Ok, removed `BertTokenizer` in favor of `BertTokenizerFast`.
